File: java/src/main/java/com/penapereira/patterns/prototype/ConcretePrototype1.java
package com.penapereira.patterns.prototype;

/** Copies its own state through a copy constructor (Bloch, Item 13). */
public final class ConcretePrototype1 implements Prototype {

    private String state;

    public ConcretePrototype1(String state) {
        this.state = state;
    }

    @Override
    public ConcretePrototype1 clone() {
        return new ConcretePrototype1(state);
    }

    public void setState(String state) {
        this.state = state;
    }

    @Override
    public String describe() {
        return "ConcretePrototype1(state=" + state + ")";
    }
}

File: typescript/src/prototype/concrete-prototype1.ts
import type { Prototype } from "./prototype.ts";

/** Copies its own state through the constructor. */
export class ConcretePrototype1 implements Prototype {
  constructor(private state: string) {}

  clone(): ConcretePrototype1 {
    return new ConcretePrototype1(this.state);
  }

  setState(state: string): void {
    this.state = state;
  }

  describe(): string {
    return `ConcretePrototype1(state=${this.state})`;
  }
}
